### DISCUSSION TITLE
Prusa MK4 + Bed Visualizer

### DISCUSSION THREAD

**Discussion by geekstar33** [jneilliii/OctoPrint-BedLevelVisualizer#619] [Q&A] [answered]
Hi, I have used your Bed Visualizer with my MK3s... now I have an MK4 and I'm curious if it will work with it?  I'm hesitant to simply press the button and find out.  I dont want to crash the nozzle or do something else bad.

Please let me know if it will capture the bed level data from an MK4.
thank you,
LM

**Comment by jneilliii** [comment]
Don't have one so can't confirm or deny. As long as you use the proper gcode for your printer for leveling there should be no issue.

**Comment by geekstar33** [comment]
Ok... MK4 uses G29.  So I changed the collection tab commands to:

@BEDLEVELVISUALIZER
G29 P1

It runs the probe sequence across the bed, but never reports anything graphically.  It just continues to say, "Please wait, retrieving current mesh".  I have to hit cancel to break out of it.  Any ideas, what I'm missing?

**Comment by jneilliii** [comment]
Those command look suspiciously similar to UBL. Try the command `M420 V` in terminal command.

**Reply by geekstar33** [reply]
Thank you so much for your responses!

I updated the collection button to this:
@BEDLEVELVISUALIZER	; instruct plugin to start recording responses from printer.
G29 P1
@BEDLEVELVISUALIZER
M420 V

Watching the terminal...
It does each probe point successfully reporting "clean and ok" with each touch point.
At the end of probing, it displays a chart of numbers in the terminal.
But at the end of the chart numbers it says this...

Recv: Mesh is invalid
Recv: Storage slot: 0
Recv: echo:Invalid mesh.
Recv: echo:Bed Leveling OFF
Recv: echo:Fade Height OFF
Recv: ok

On the Bed Visualizer tab, it does not update the graphics and still says, "Please wait, retreiving current mesh"... but never completes.

Hoping to solve this for other MK4 users too....

**Reply by geekstar33** [reply, marked as answer]
**UPDATE:**

Ok... I got something... I looked at what the Prusa Slicer was using in the start gcode:  here is the updated collection code I used...

@BEDLEVELVISUALIZER	; instruct plugin to start recording responses from printer.
G29 P1
G29 P3.2 ; interpolate mbl probes
G29 P3.13 ; extrapolate mbl outside probe area
G29 A ; activate mbl
M420 V

the graphics were rendered... but wow! It gave back a lot of points and the MK4 suprisingly had a large variance.  Although MK4 does a great job with bed leveling... its still nice to see what is going on with the bed.

I believe it needed the G29 P3.2 to interpolate and make a valid mesh.

![MK4 probe](https://github.com/jneilliii/OctoPrint-BedLevelVisualizer/assets/137321642/bd346046-47c9-4987-b1e7-65f8bf3c6120)

Hopefully this will help others.  Feedback still appreciated.

**Comment by jneilliii** [comment]
You should move the command `@BEDLEVELVISUALIZER` just before the M420 V command I think.

**Comment by mt-mrx** [comment]
Thanks for this post. With the following instructions I get a proper visualization.
```
; Prusa MK4
G29 P1
G29 P3.2 ; interpolate mbl probes
G29 P3.13 ; extrapolate mbl outside probe area
G29 A ; activate mbl
@BEDLEVELVISUALIZER ; instruct plugin to start recording responses from printer.
M420 V
```

![Selection_096](https://github.com/jneilliii/OctoPrint-BedLevelVisualizer/assets/64284703/da5be819-7b76-46f8-9f8f-dfd33db78d1e)
